core/src/main/java/org/apache/calcite/rel/RelDistributions.java

```diff
@@ -135,10 +141,14 @@ public RelDistribution apply(Mappings.TargetMapping mapping) {
       if (keys.isEmpty()) {
         return this;
       }
-      return getTraitDef().canonize(
-          new RelDistributionImpl(type,
-              ImmutableIntList.copyOf(
-                  Mappings.apply2((Mapping) mapping, keys))));
+      for (int key : keys) {
+        if (mapping.getTargetOpt(key) == -1) {
```

Comment:
Key might be greater than / equal with mapping.sourceCount (out of range). Haven't try to reproduce here in Calcite. Perhaps we can also add a check in TargetMapping implementation class, if the key >= sourceCount, also return -1.

Reply:
I agree. Added fix for mappings.